src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/BehaviourService.kt

```diff
@@ -0,0 +1,69 @@
+package uk.gov.justice.digital.hmpps.incentivesapi.service
+
+import kotlinx.coroutines.flow.toList
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.incentivesapi.dto.prisonapi.PrisonerCaseNoteByTypeSubType
+import uk.gov.justice.digital.hmpps.incentivesapi.jpa.PrisonerIepLevel
+import java.time.Clock
+import java.time.LocalDateTime
+import java.time.Period
+
+@Service
+class BehaviourService(
+  private val prisonApiService: PrisonApiService,
+  private val clock: Clock,
+) {
+  private fun behaviourCaseNoteMap() = mapOf("POS" to "IEP_ENC", "NEG" to "IEP_WARN")
+
+  suspend fun getBehaviours(reviews: List<PrisonerIepLevel>) =
+    getCaseNoteUsageByLastReviewDate(
+      prisonApiService.retrieveCaseNoteCountsByFromDate(
+        behaviourCaseNoteMap().keys.toList(),
+        getLastRealReviewForOffenders(reviews)
+      ).toList(),
+      behaviourCaseNoteMap(),
+    )
+  private fun getCaseNoteUsageByLastReviewDate(caseNotesByType: List<PrisonerCaseNoteByTypeSubType>, subTypeMap: Map<String, String>) =
+    caseNotesByType
+      .groupBy(PrisonerCaseNoteByTypeSubType::toKey)
+      .map { cn ->
+        CaseNoteSummary(
+          key = cn.key,
+          totalCaseNotes = calcTypeCount(cn.value),
+          numSubTypeCount = calcTypeCount(cn.value.filter { cnc -> cnc.caseNoteSubType == subTypeMap[cn.key.caseNoteType] })
+
+        )
+      }.associateBy { it.key }
+
+  private fun calcTypeCount(caseNoteUsage: List<PrisonerCaseNoteByTypeSubType>): Int =
+    caseNoteUsage.map { it.numCaseNotes }.fold(0) { acc, next -> acc + next }
+
+  private fun getLastRealReviewForOffenders(reviews: List<PrisonerIepLevel>): Map<Long, LocalDateTime> =
+    reviews
+      .groupBy { it.bookingId }
+      .map { review ->
+        val latestReview = review.value.firstOrNull(PrisonerIepLevel::isRealReview) ?: review.value.first()
```

Comment:
Are we sure this `?: review.value.first()` is the correct thing to do?

Someone may have been in prison for months/years without a genuine Incentive review.
They may have a number of relevant case notes in the last 3 months.
But if, let's say, they got transferred last week this `review.value.first()` would mean only case notes from the last week would returned, right?

Maybe this should fallback to 3 months ago when no real review is found? Sorry if this is a stupid question, I'm still working my way through this PR.

Reply:
Its a good point. I could change it so that if no real review found then just get three months

Reply:
@xoen I've updated this to your suggestion